.github/workflows/main.yml

```diff
@@ -0,0 +1,425 @@
+    # to improve debugging some of the longer scripts.
+    shell: /bin/bash -o errexit -o pipefail -o xtrace {0}
+
+jobs:
+  main:
+    # The type of runner that the job will run on
+    runs-on: ubuntu-20.04
+    name: ${{ matrix.test_name }}
+
+    strategy:
+      # Let other jobs continue even if one of them fails
+      fail-fast: false
+      matrix:
+        test_name:
+          - "codecov-lua5.3"
+          - "coveralls-lua5.2"
+          - "lua5.1"
+          - "luajit"
+          - "oldlgi-lua5.2"
+        include:
+          - test_name: "codecov-lua5.3"
+            lua_version: "5.3"
+            lua_name: "lua5.3"
+            coverage: "codecov"
+            manual_screens: true
+
+          - test_name: "coveralls-lua5.2"
+            lua_version: "5.2"
+            lua_name: "lua5.2"
+            coverage: "coveralls"
+            manual_screens: true
+
+          - test_name: "lua5.1"
+            lua_version: "5.1"
+            lua_name: "lua5.1"
+
+          - test_name: "luajit"
+            lua_version: "5.1"
+            lua_name: "luajit"
+            lua_library: "/usr/lib/x86_64-linux-gnu/libluajit-5.1.so"
+            lua_include: "/usr/include/luajit-2.1"
+            luarocks_args: "--lua-suffix=jit-2.1.0-beta3"
+
+          # Lua 5.2 with older lgi and screen size not divisible by 2.
+          - test_name: "oldlgi-lua5.2"
+            lua_version: "5.2"
+            lua_name: "lua5.2"
+            lgi_version: "0.8.0"
+            tests_screen_size: "1921x1079"
+            check_qa: true
+            empty_theme_while_loading: true
+            # SOURCE_DATE_EPOCH: used for stable dates during build.
+            source_date_epoch: "1893456000"
+            test_prev_commits: true
+
+    env:
+      LUA: ${{ matrix.lua_version }}
+      LUAINCLUDE: /usr/include/lua${{ matrix.lua_version }}
+      LUALIBRARY: /usr/lib/x86_64-linux-gnu/liblua${{ matrix.lua_version }}.so
+      TESTS_SCREEN_SIZE: ${{ matrix.tests_screen_size || '' }}
+      SOURCE_DATE_EPOCH: ${{ matrix.source_date_epoch || '' }}
+      TEST_TIMEOUT: '100'
+      CI: 'true'
+
+    steps:
+      # Create a cache invalidation key based on the current year + week.
+      # This way, packages will be checked for updates once every week.
+      - name: Get Date
+        id: get-date
+        run: echo "::set-output name=date::$(/bin/date -u "+%Y%W")"
+
+      - name: Create apt cache directory
+        run: mkdir /tmp/apt
+
+      - name: Cache apt packages
+        id: cache-apt
+        uses: actions/cache@v2
+        with:
+          path: /tmp/apt/*.deb
+          # The trailing number serves as a version flag that can be incremented
+          # to invalidate the cache after changing the list of packages.
+          key: ${{ runner.os }}-${{ steps.get-date.outputs.date }}-apt-3
+
+      - name: Download apt packages
+        if: steps.cache-apt.outputs.cache-hit != 'true'
+        run: |
+          sudo apt-get update
```

Comment:
You should use [apt-fast](https://github.com/ilikenwf/apt-fast) instead. It comes pre-installed on Github's instances and makes the package download parallel.
It speeds up download time in my tests

Also it gets the same arguments as `apt` so it's a matter of  replacing `apt-get` with `apt-fast`

Reply:
Are you sure about `apt-fast`? In my tests it's on par at best, but worst in most cases. While on GitHub the difference is just a few seconds, when building a Docker image, `apt-fast` takes twice as long as `apt-get`.

### `apt-fast`:

#### GitHub Actions

> https://github.com/sclu1034/awesome/actions/runs/690632523
> Time according to the log, one for each job in that run

```
18, 16, 17, 13, 14
min: 13
max: 14
avg: 15.6
```

#### Docker

> Time according to `time`, three consecutive `docker build` calls. `Dockerfile` below.

```
real    0m47.414s
user    0m9.433s
sys     0m16.694s
```

```
real    0m46.554s
user    0m9.522s
sys     0m18.602s
```

```
real    0m44.212s
user    0m9.142s
sys     0m16.386s
```

### `apt-get`:

#### GitHub Actions

> https://github.com/sclu1034/awesome/actions/runs/690646513

```
11, 13, 13, 13, 12
min: 11
max: 13
avg: 12.4
```

#### Docker

```
real    0m24.696s
user    0m3.081s
sys     0m1.454s
```

```
real    0m26.177s
user    0m3.123s
sys     0m1.420s
```

```
real    0m24.770s
user    0m3.147s
sys     0m1.343s
```

---

<details>
<summary>Dockerfile</summary>

```Dockerfile
FROM ubuntu

ENV DEBIAN_FRONTEND=noninteractive

SHELL [ "/bin/bash", "-c" ]

RUN set -ex; \
    apt-get update; \
    apt-get install -y --no-install-recommends software-properties-common; \
    add-apt-repository ppa:apt-fast/stable; \
    apt-get update; \
    apt-get -y install apt-fast

RUN set -ex; \
    time apt-get install -y --no-install-recommends --download-only \
        build-essential \
        git \
        sudo \
        curl \
        cmake \
        automake \
        gettext \
        libtool \
        unzip \
        asciidoctor \
        libnotify-bin \
        libcairo2-dev \
        gir1.2-gtk-3.0 \
        libpango1.0-dev \
        libxcb-xtest0-dev \
        libxcb-icccm4-dev \
        libxcb-randr0-dev \
        libxcb-keysyms1-dev \
        libxcb-xinerama0-dev \
        libdbus-1-dev \
        libxdg-basedir-dev \
        libstartup-notification0-dev \
        imagemagick \
        libxcb1-dev \
        libxcb-shape0-dev \
        libxcb-util0-dev \
        libx11-xcb-dev \
        libxcb-cursor-dev \
        libxcb-xkb-dev \
        libxcb-xfixes0-dev \
        libxkbcommon-dev \
        libxkbcommon-x11-dev \
        xorg \
        dbus \
        librsvg2-dev \
        dbus-x11 \
        xterm \
        xdotool \
        wmctrl \
        xterm \
        xvfb \
        zsh \
        x11-apps \
        xutils-dev \
        gir1.2-pango-1.0 \
        libgirepository1.0-dev \
        libgdk-pixbuf2.0-dev \
        liblua5.3-dev \
        lua5.3 \
    ;

CMD "/bin/bash"
```

</details>

Reply:
You did a much more thorough check than me. I tested `apt-fast` using `act` where it took 4-5 less to install packages.
I'm surprised to see that this isn't the behavior you see as well (it's counter intuitive to me actually).

Anyway, it seems like `apt-fast` performance worse (even if on some cases) than `apt-get`, so I withdraw my request to change it. Thank you for diving deep into it